### PR TITLE
Issue #14, #18, #19

### PR DESCRIPTION
**Pull request by cw646**
Included maps for vertices and elements, and method to use vertices to generate points within an element using Eigen library.

### CODE REVIEW

File: include/gmsh.h [hunk 1]
```diff
@@ -5,12 +5,17 @@
...
+#include <map>
 #include <memory>
 #include <vector>
-
 #include "mesh.h"
 #include "point.h"
 
+//libraries
+#include <Eigen/Dense>
+using Eigen::MatrixXd;
```
Comment (kks32) [inline include/gmsh.h]: Avoid shortening of namespace, use `Eigen::MatrixXd` everywhere to avoid namespace clash.

File: include/gmsh.h [hunk 2]
```diff
@@ -21,17 +26,26 @@ class GMSH : public Mesh {
   //! Read elements in GMSH
   void read_elements(const std::string& filename);
 
+  //!Store element id and vertices coordinates as map
```
Comment (kks32) [inline include/gmsh.h]: You need a space after `//!`.

File: include/gmsh.h [hunk 3]
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
...
+  std::vector<std::shared_ptr<Point<3>>> materialpoints_;
 
   //! Stress vector in Voigt Notation
-  //! $\sigma_{xx}$ $\sigma_{yy}$ $\sigma_{zz}$ $\tau_{yz}$ $\tau_{zx}$ $\tau_{xy}$
+  //! $\sigma_{xx}$ $\sigma_{yy}$ $\sigma_{zz}$ $\tau_{yz}$ $\tau_{zx}$
+  //! $\tau_{xy}$
   std::vector<std::array<double, 6>> stress_;
+
+  //! Map to store id and vertices coordinates
+  std::map<double, std::array<double, 3>> vertices_;
```
Comment (kks32) [inline include/gmsh.h:65]: Why is the `id` a double instead of an unsigned?

File: include/gmsh.h [hunk 4]
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
...
+
+  //! Map to store element ID and vertices ID
+  // 4 = element vertices
+  std::map<double, std::array<double, 4>> elements_;
```
Comment (kks32) [inline include/gmsh.h]: ids need to be unsigned

File: include/gmsh.tcc [hunk 5]
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
         istream >> elementry;
 
         //! \brief Check element type
-        //! \details If element type not == to Tdim, skip element
-        if (elementtype != ndimension) {
+        //! \details If element type not == to specified elementtype_, skip
+        //! element
+        if (elementtype != elementype) {
```
Comment (kks32) [inline include/gmsh.tcc]: Is this right? Should it be `elementtype != elementtype_`

File: include/gmsh.tcc [hunk 6]
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
...
+            verticesarray[10] = verticesfind->second[1];
+            verticesarray[11] = verticesfind->second[2];
+          } else {
+          }
+        } else {
+        }
+      }
+      elementcoordinates_.insert(
+          std::make_pair(elementfind->first, verticesarray));
+    } else {
```
Comment (kks32) [inline include/gmsh.tcc]: Why an empty else?

File: include/gmsh.tcc [hunk 7]
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
...
+      elementkeyvalues[3] = elementfind->second[3];
+
+      for (unsigned q = 0; q < 4; ++q) {
+        if (q == 0) {
+          verticesfind = vertices_.find(elementkeyvalues[0]);
+          if (elementfind != elements_.end()) {
+            verticesarray[0] = verticesfind->second[0];
+            verticesarray[1] = verticesfind->second[1];
+            verticesarray[2] = verticesfind->second[2];
+          } else {
```
Comment (kks32) [inline include/gmsh.tcc]: Remove empty elses.

File: include/gmsh.tcc [hunk 8]
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
...
+  for (unsigned q = firstelement; q < lastelement + 1; ++q) {
+    elementfind = elements_.find(q);
+    if (elementfind != elements_.end()) {
+
+      elementkeyvalues[0] = elementfind->second[0];
+      elementkeyvalues[1] = elementfind->second[1];
+      elementkeyvalues[2] = elementfind->second[2];
+      elementkeyvalues[3] = elementfind->second[3];
+
+      for (unsigned q = 0; q < 4; ++q) {
```
Comment (kks32) [inline include/gmsh.tcc]: What's the 4, 12, make them `const`s

File: include/gmsh.tcc [hunk 9]
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
...
+      elementcoordinates_.insert(
+          std::make_pair(elementfind->first, verticesarray));
+    } else {
+    }
+  }
+}
+
+void GMSH::compute_material_points() {
+
+  double xcoord;
```
Comment (kks32) [inline include/gmsh.tcc]: Can you not use a 3D array?

File: include/gmsh.tcc [hunk 10]
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
...
+  double ycoord;
+  double zcoord;
+
+  std::array<double, 3> pointsarray;
+  std::map<double, std::array<double, 12>>::iterator coordinatesfind;
```
Comment (kks32) [inline include/gmsh.tcc]: What's 12?

File: include/gmsh.tcc [hunk 11]
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
...
+      m(1, 2) = coordinatesfind->second[7];
+      m(2, 2) = coordinatesfind->second[8];
+
+      m(0, 3) = coordinatesfind->second[9];
+      m(1, 3) = coordinatesfind->second[10];
+      m(2, 3) = coordinatesfind->second[11];
+
+      // Centroid test
+      //++++++++++++++++++++++++++++++++++++++
+      xcoord = 0.25 * (m(0, 0) + m(0, 1) + m(0, 2) + m(0, 3));
```
Comment (kks32) [inline include/gmsh.tcc]: Write it directly to `pointsarray`.

Comment (kks32) [inline include/gmsh.tcc]: Can you not put this inside a loop?

File: include/gmsh.tcc [hunk 12]
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
...
+  double ycoord;
+  double zcoord;
+
+  std::array<double, 3> pointsarray;
+  std::map<double, std::array<double, 12>>::iterator coordinatesfind;
+
+  const unsigned firstelementcoord = elementcoordinates_.begin()->first;
+  const unsigned lastelementcoord = elementcoordinates_.rbegin()->first;
+
+  for (unsigned t = firstelementcoord; t < lastelementcoord + 1; ++t) {
```
Comment (kks32) [inline include/gmsh.tcc:191]: We need to rethink this algorithm. Is this through all the element vertices?

File: src/main.cc [hunk 13]
```diff
@@ -18,12 +18,14 @@ int main(int argc, char** argv) {
     //! Mesh handler
     std::unique_ptr<GMSH> mesh(new GMSH());
 
-    //! Read mesh and compute stresses
+    //! Read mesh and compute material points & stresses
     mesh->read_vertices(io->mesh_file_name());
+    mesh->read_elements(io->mesh_file_name());
```
Comment (kks32) [inline src/main.cc:23]: Read elements should call `store_elements` at the end of the function, no reason to call another function in main.

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/io.tcc [hunk 14]
```diff
@@ -1,21 +1,21 @@
...
     //! Write the total number of vertices
-    vertices_file << vertices.size() << "\n";
+    vertices_file << materialpoints.size() << "\n";
 
     //! Write the coordinates of the vertices
     //! [X] [Y] [Z]
     //! Note that for 2D, z values are 0
     //! For 1D, both y and z values are 0
-    for (auto const& vertex : vertices) {
+    for (auto const& vertex : materialpoints) {
```
Comment (ezrayst) [inline include/io.tcc]: is it better to use `point` instead of `vertex` here?

File: include/io.tcc [hunk 15]
```diff
@@ -24,12 +24,13 @@ void IO::write_vertices(
...
   }
 
-  std::cout << "Wrote material points id and coordinates\n";
+  std::cout << "Wrote material point coordinates\n";
 }
 
 //! \brief Write initial stresses of material points
 //! \param[in[ stresses Initial stress of material points
 void IO::write_stresses(const std::vector<std::array<double, 6>>& stresses) {
+  double id = 1;
```
Comment (ezrayst) [inline include/io.tcc]: any specific reason you used `double` instead `unsigned`

File: include/io.tcc [hunk 16]
```diff
@@ -40,14 +41,17 @@ void IO::write_stresses(const std::vector<std::array<double, 6>>& stresses) {
...
     //! Stresses in Voigt Notation
+    // Stress needs ID too
     //! $\sigma_{xx}$ $\sigma_{yy}$ $\sigma_{zz}$
     //! $\tau_{yz}$ $\tau_{zx}$ $\tau_{xy}$
     for (const auto& stress : stresses) {
       stress_file.setf(std::ios::fixed, std::ios::floatfield);
+      stress_file << id << '\t';
       for (double stress_component : stress) {
-        stress_file << stress_component << '\t';
+        stress_file << 0 << '\t';
```
Comment (ezrayst) [inline include/io.tcc]: I will handle this part

File: include/gmsh.tcc [hunk 17]
```diff
@@ -66,9 +64,13 @@ void GMSH::read_elements(const std::string& filename) {
 
   const unsigned toplines = 4;
   const unsigned ndimension = 3;
+  // specify element type 4 = tetrahedral
```
Comment (ezrayst) [inline include/gmsh.tcc:67]: just wondering. isn't it `quadrilateral?`

Comment (cw646) [reply include/gmsh.tcc:67]: For unstructured 2D it's triangles, and for unstructured 3D it's 4 node-tetrahedral.

File: include/gmsh.tcc (hunk 5 above)
Comment (cw646) [reply include/gmsh.tcc]: Ah yes, I removed the trailing _ in previous commit, but should rename it to something else.

**Pull request closed without merging**